tracer/src/Datadog.Trace/DataStreamsMonitoring/Aggregation/DataStreamsAggregator.cs

```diff
@@ -46,26 +53,56 @@ public DataStreamsAggregator(DataStreamsMessagePackFormatter formatter, int buck
     /// </summary>
     public void Add(in StatsPoint point)
     {
-        var currentBucketStartTime = BucketStartTimeForTimestamp(point.TimestampNs);
-        AddToBuckets(point, currentBucketStartTime, _currentBuckets);
+        try
+        {
+            _lock.TryEnterReadLock(_lockTimeoutMs);
+            try
+            {
+                var currentBucketStartTime = BucketStartTimeForTimestamp(point.TimestampNs);
+                AddToBuckets(point, currentBucketStartTime, _currentBuckets);
 
-        var originTimestamp = point.TimestampNs - point.PathwayLatencyNs;
-        var originBucketStartTime = BucketStartTimeForTimestamp(originTimestamp);
-        AddToBuckets(point, originBucketStartTime, _originBuckets);
+                var originTimestamp = point.TimestampNs - point.PathwayLatencyNs;
+                var originBucketStartTime = BucketStartTimeForTimestamp(originTimestamp);
+                AddToBuckets(point, originBucketStartTime, _originBuckets);
```

Comment:
As you know this is not thread safe (multiple threads can enter in read mode/lock) and you have code to modify the bucket dictionary in AddToBuckets, even this may not throw an exception you could be dropping points without noticing it or incrementing the counter. Are you ok with that?

Modifying dictionaries on a read lock makes this code confusing to read.
I guess the DDSketch object is thread safe, consider enabling a write lock when modifying the dictionary.